File: 0309_best_time_to_buy_sell_cooldown/max_profit.py
from typing import List


def max_profit(prices: List[int]) -> int:
    """
    Time: O(n)
    Space: O(1)
    """
    notHold, notHold_cooldown, hold = 0, float('-inf'), float('-inf')
    for p in prices:
        hold = max(hold, notHold - p)
        notHold = max(notHold, notHold_cooldown)
        notHold_cooldown = hold + p
    return max(notHold, notHold_cooldown)

File: 0309_best_time_to_buy_sell_cooldown/max_profit.go
package main

import "math"

// Time: O(n)
// Space: O(1)
func maxProfit(prices []int) int {
	hold, notHold, notHoldCoolDown := math.MinInt16, 0, math.MinInt16
	for _, p := range prices {
		hold = max(hold, notHold-p)
		notHold = max(notHold, notHoldCoolDown)
		notHoldCoolDown = hold + p
	}
	return max(notHold, notHoldCoolDown)
}

func max(x, y int) int {
	if x >= y {
		return x
	}
	return y
}

func main() {

}
